pkg/collections.go

```diff
@@ -0,0 +1,308 @@
+//
+// DISCLAIMER
+//
+// Copyright 2020 ArangoDB GmbH, Cologne, Germany
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Copyright holder is ArangoDB GmbH, Cologne, Germany
+//
+// Author Gergely Brautigam
+//
+package pkg
+
+import (
+	"context"
+	"errors"
+
+	"github.com/arangodb/go-driver"
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sync/semaphore"
+)
+
+// copyCollections copies all collections for a database.
+func (c *copier) copyCollections(ctx context.Context, db driver.Database) error {
+	log := c.Logger
+	log.Info().Msg("Beginning to copy over collection data.")
+	var (
+		sourceDB      driver.Database
+		collections   []driver.Collection
+		destinationDB driver.Database
+	)
+	if err := c.backoffCall(ctx, func() error {
+		sdb, err := c.sourceClient.Database(ctx, db.Name())
+		if err != nil {
+			return err
+		}
+		sourceDB = sdb
+		return nil
+	}); err != nil {
+		return err
+	}
+	if err := c.backoffCall(ctx, func() error {
+		colls, err := sourceDB.Collections(ctx)
+		if err != nil {
+			c.Logger.Error().Err(err).Msg("Failed to list collections for source database.")
+			return err
+		}
+		collections = colls
+		return nil
+	}); err != nil {
+		return err
+	}
+	if err := c.backoffCall(ctx, func() error {
+		ddb, err := c.destinationClient.Database(ctx, sourceDB.Name())
+		if err != nil {
+			c.Logger.Error().Err(err).Msg("Failed to get destination database.")
+			return nil
+		}
+		destinationDB = ddb
+		return nil
+	}); err != nil {
+		return err
+	}
+
```

Comment:
We should sort the collections list here in the following way:
  - collections without `distributeShardsLike` should go before those with the attribute
  - vertex collections (non-edge collections) should go before edge collections
Then we should **create** the collections sequentially in this order at the destination.
Then we should copy the **data** in parallel as it is done currently.
This should ensure that we can transport the `distributeShardsLike` relationship faithfully and without failures. 
Maybe we need yet another trick for smart graphs, I would have to check this or ask Michael.

Reply:
Reason is: if A is a collection and B has `distributeShardsLike` A, then B can only created when A is there. But `distributeShardsLlike` can only point to a collection **without** `distributeShardsLike` attribute.